Fix double-forward, prefer legacy forward maps

valentinewallace opened the pull request:
Addresses a chunk of the feedback from https://github.com/lightningdevkit/rust-lightning/pull/4227#pullrequestreview-3568348748 (tracked in #4280). Splitting it out for ease of review. Partially addresses #4286. 

  - Fix a bug that would cause double-forwarding of inbound HTLCs when using reconstructed forward maps
  - Prefer legacy forward maps in production while randomly using reconstructed maps in tests for coverage (see commit message) 
  - a few other nits from the aforementioned review

valentinewallace commented:
Going to take another look at this tomorrow before un-drafting it

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -16681,7 +16685,18 @@ where
+		#[cfg(not(any(test, feature = "_test_utils")))]
+		let reconstruct_manager_from_monitors = false;
+		#[cfg(any(test, feature = "_test_utils"))]
+		let reconstruct_manager_from_monitors = true;
```
Plan looks good to me. Only the following q:

Why do we first want to merge all the desired reconstruction logic? Is that to avoid multiple "reconstruct_from_monitors" flags? If that's the reason, perhaps it is worth to have multiple flags, so that we can skip stage 1 and keep the various reconstruction changes independent.

valentinewallace replied:
Went with the randomizing option, it simplified the code further and seems to achieve what we want.

TheBlueMatt commented on lightning/src/ln/reload_tests.rs:
```diff
@@ -1258,6 +1258,62 @@ fn do_manager_persisted_pre_outbound_edge_forward(intercept_htlc: bool) {
 }
 
+#[test]
+fn test_manager_persisted_post_outbound_edge_forward() {
```
Please add a comment describing at a high level why we wrote this test and what the intent of the test is.

valentinewallace commented:
Addressed feedback, main diff is [here](https://github.com/lightningdevkit/rust-lightning/compare/425747e140d297d4bca229895c93fc6ac3f913c5..b7ab495a5dc5b8dce4854aa49e7bed4f8a9f2ad0). Also pushed some whitespace fixes after.

valentinewallace commented:
Discussed offline, going to add an environmental variable to set which manager reconstruction paths to use. Rebased on main to get the changes from #4296

valentinewallace commented:
Added the environment variable: [diff](https://github.com/lightningdevkit/rust-lightning/compare/f1ea7bb7c6a92cebec5bcd205978ac55c1cfc657..f245139dfb86185f7bab6c2fcda6a6032708132c)

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -18024,12 +18034,6 @@ where
 							// still have an entry for this HTLC in `forward_htlcs`,
 							// `pending_intercepted_htlcs`, or `decode_update_add_htlcs`, we were apparently not
```
decode_update_add_htlcs -> decode_update_add_htlcs_legacy

joostjager replied:
Perhaps explain here why the legacy adds only need to be cleaned for closed channels, and the reconstructed ones always?

valentinewallace replied:
I believe for legacy adds it's explained decently in the error messages in the calls to `dedup_decode_update_add_htlcs`, and the reconstructed reasoning is explained in a comment. Let me know if you want some specific phrasing/what is missing.

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11747,6 +11747,10 @@ This indicates a bug inside LDK. Please report this error at https://github.com/
+				// It's possible we processed this intercept forward, generated an event, then re-processed
+				// it here after restart, in which case the intercept event should not be pushed
+				// redundantly.
+				new_intercept_events.retain(|ev| !events.contains(ev));
```
Verified that this is covered by the test suite. (`manager_persisted_pre_outbound_edge_intercept_forward`)

joostjager commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -17930,6 +17934,36 @@ where
+		// persist that state, relying on it being up-to-date on restart. Newer versions are moving
+		// towards reducing this reliance on regular persistence of the `ChannelManager`, and instead
+		// reconstruct HTLC/payment state based on `Channel{Monitor}` data if
+		// `reconstruct_manager_from_monitors` is set below. Currently it is only set in tests, randomly
```
Bringing everything together, would the upgrade/downgrade situation look like this?

Version|Read|Write | Upgrade to version (max) | Downgrade to version (min)
-|-|-|-|-
0.2 | legacy | legacy | 0.5 |  -
0.3 | legacy | legacy+new | 0.6 | 0.2
0.4 | legacy+new | legacy+new | 0.6 | 0.2
0.5 | legacy+new | new | 0.6 | 0.4
0.6 | new | new | - | 0.4

Also wondering - in relation to the discussion in yesterday's sync meet - how the `reconstruct_manager_from_monitors` makes for a faster path?

TheBlueMatt replied:
> would the upgrade/downgrade situation look like this?

My understanding is that our goal is that 0.3 would support reading new as well, so that 0.5 can downgrade to 0.3 rather than 0.4.

> Also wondering - in relation to the discussion in yesterday's sync meet - how the reconstruct_manager_from_monitors makes for a faster path?

So that we get the above. I don't see a reason to want to only allow 0.5 to downgrade to 0.4 rather than 0.3. The code currently doesn't do that but presumably in a followup we could do that?

joostjager replied:
@valentinewallace and I discussed this more. It seems that having a dedicated flag for signaling that the old maps are not written is better than guessing at a future version number where this is the case.

Version|Read|Write | Upgrade to version (max) | Downgrade to version (min) | Set flag
-|-|-|-|-|-
0.2 | legacy | legacy | 0.5 |  - | 
0.3 | legacy OR new with flag | legacy+new | 0.6 | 0.2| 
0.5 | legacy OR new with flag | new | 0.6 | 0.3 | X
0.6 | new | new | - | 0.3| X

TheBlueMatt replied:
> It seems that having a dedicated flag for signaling that the old maps are not written is better than guessing at a future version number where this is the case.

You mean instead of using `SERIALIZATION_VERSION`/`MIN_SERIALIZATION_VERSION` constants you want to use a TLV? I guess that's fine, but it seems much simpler to use the version numbers so that we can also drop some of the legacy crap that is written as non-TLVs that we'll never be writing anymore. Don't really see a reason to avoid that.

The upgrade path looks right to me, though.

joostjager replied:
Would that look like this? 

Version|Read|Write | Upgrade to version (max) | Downgrade to version (min) | SER_VER | MIN_SER_VER
-|-|-|-|-|-|-
0.2 (current) | legacy | legacy | 0.4 |  any | 1 | 1 
0.3 | legacy OR new with flag | legacy+new | any | any | 2 | 1 
0.4 | legacy OR new with flag | legacy+new | any | any | 2 | 1 
0.5 | new | new | any | 0.3 | 3 | 2

We are wondering what happens if we need to bump in between for a different reason.

There is also the issue that the new format is really a per-htlc format change, not overall chan mgr version.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -18020,17 +18056,20 @@ where
+							// HTLCs to `decode_update_add_htlcs` in the above loop, but we need to prune from
+							// those added HTLCs if they were already forwarded to the outbound edge. Otherwise,
+							// we'll double-forward.
+							if reconstruct_manager_from_monitors {
```
Hmm, I personally find it more readable to be explicit since we only want to dedup that map if `reconstruct_manager_from_monitors` is set. But I believe you're correct.
